### ISSUE TITLE
Auto Snapshot for Last N Deployments

### ISSUE THREAD

**Issue by MarianRaphael** [FlowFuse/flowfuse#2423]
### Epic

https://github.com/flowforge/flowforge/issues/2421

### User Story

As a FlowForge User,

I want to have the possibility to automatically create snapshots of my last 'n' NR deployments,

So that I can recover quickly and easily when there is a crash or misconfiguration.

### Which customers would this be availble to

Licensed Edition (EE)
Teams Tier

### Acceptance Criteria

- [ ] should provide an option for users to set the number 'n' they want to back up (with a cap)
- [ ] When this option is activated, FF should automatically create snapshots of the NR Instance and Device when new deployed
- [ ] possibility for Instances
- [ ] possibility for Devices

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate
```[tasklist]
### Instances
- [ ] https://github.com/FlowFuse/flowfuse/issues/3355
```

```[tasklist]
### Devices
- [ ] https://github.com/FlowFuse/flowfuse/issues/3357
- [ ] https://github.com/FlowFuse/flowfuse/issues/3358
```

**Comment by knolleary** [comment]
Only just saw this has been added to the development plan for this release. We need to flush out the requirements in order to get something into this release without a lot of time to do so.

## Proposal

### Configuration

 - For Instances we can add to the Template object
 - For Devices - we have nowhere to store device specific configuration like this. Don't have a quick solution for that.

### Managing snapshots

The snapshot does not currently identify how it was created. If we are going to delete the N+1 auto-generated snapshot, we need to make sure that isn't a manually generated snapshot the user actually wants to keep.

The only mechanism we have for that is the name of the snapshot.

 - Auto-generated snapshots have `[auto]` prepended to their name. We can use that to ensure at most `N` snapshots exist that match that pattern.
 - If a user wants to save an auto-generated snapshot, they edit the name to remove the `[auto]` part.

### When to create the snapshot

We need to identify how/when to generate the snapshot - what triggers it exactly. Typically that will be when the user hits the deploy button and the flows are sent to the storage api in the platform.

However, that will also be triggered when a user runs a pipeline, or reverts to a previous snapshot. In those cases, we already have the snapshot and it would feel odd to auto generate another one. But the call to the storage api doesn't have the context of what triggered it.

I'm not immediately sure how we distinguish between the two scenarios.

### Devices

With the device editor there is no platform API that gets called when the deploy button is hit - that is all done locally on the device. So triggering an auto-snapshot would have to be done on the device itself. That implies the device needs to know what auto-snapshot configuration is in place and have it post the snapshot itself. Even then - we don't currently have any hooks in the device to know the deploy button has been hit. There's quite a lot of development work needed to make this work on devices.

**Comment by Steve-Mcl** [comment]
#### Other points not clear to me:

* Specifying "Last N Deployments" infers: 
   * there is automatic clean up? We would need to be careful not to clean up a snapshot in use (see existing issue about deleting snapshot resets device) _[@Steve-Mcl note-to-self, update this with issue number]_
   * a snapshot each time a deploy is made? This could very easily and very quickly "use up" all `N` snapshots on a busy day with a cautious user deploying after every little change. Some thought into the exact timing is needed. (e.g. snapshot once on first deploy but not again for x hours?  Never overwrite N+1 unless it is > 48h? etc etc)

#### Musings
Perhaps `N` should not a number but instead, a "duration"? I.e. if we snapshot with each deploy, then these auto snapshots have a TTL of (for example) 48h then are cleaned up after `N` has passed?

**Comment by joepavitt** [comment]
Isn't this already implemented @MarianRaphael with #3355 and #3358?

**Comment by MarianRaphael** [comment]
Yes
